lib/src/cli.rs

```diff
@@ -324,26 +332,32 @@ async fn container_import(
     let opts = UnencapsulateOptions {
         progress: Some(tx_progress),
     };
-    let import = crate::container::unencapsulate(repo, imgref, Some(opts));
-    tokio::pin!(import);
-    tokio::pin!(rx_progress);
-    let import = loop {
-        tokio::select! {
-            _ = rx_progress.changed() => {
-                let n = rx_progress.borrow().processed_bytes;
+    let rx_progress_stream =
+        tokio_stream::wrappers::WatchStream::new(rx_progress).map(ProgressOrFinish::Progress);
+    let import = crate::container::unencapsulate(repo, imgref, Some(opts))
+        .into_stream()
+        .map(ProgressOrFinish::Finished);
+    let stream = rx_progress_stream.merge(import);
+    tokio::pin!(stream);
+    let mut import_result = None;
+    while let Some(value) = stream.next().await {
+        match value {
+            ProgressOrFinish::Progress(progress) => {
+                let n = progress.borrow().processed_bytes;
                 if let Some(pb) = pb.as_ref() {
                     pb.set_message(format!("Processed: {}", indicatif::HumanBytes(n)));
                 }
             }
-            import = &mut import => {
+            ProgressOrFinish::Finished(import) => {
                 if let Some(pb) = pb.as_ref() {
                     pb.finish();
                 }
-                break import?;
```

Comment:
I think that dropping this `break` may now result in `pb.set_message()` possibly being called after a `pb.finish()`, depending on the ordering of items in the merged stream. I'm not too familiar with `indicatif`, but that behavior sounds risky.

If this was an intended change with the aim of fully driving the stream to completion, then I think the `Progress => {...}` arm should also be augmented with a `if !pb.is_finished()` check.

Reply:
Great catch!  I chose to move the progress bar completion outside of the stream processing which seems simplest.

Though, it seems like this pattern where one element of the stream is a singleton could re-occur, and we could find/create a helper for this.  But...eh.

I'm still not 100% sure that the original use of `select!` is buggy...we previously did this:

````
let import = crate::container::unencapsulate(repo, imgref, Some(opts));
tokio::pin!(import);
```

which seems to me should prevent the value from being dropped and re-created.  

(A bit of time passes)

OK yes, [this linked blog entry](https://tomaka.medium.com/a-look-back-at-asynchronous-rust-d54d63934a1c#3008) has:

> Rewrite the select! to not destroy the future. [Example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d0c95a33aab6859bfda3405e9b29b5ef). This is arguably the best solution in that specific situation, but it can sometimes introduce a lot of complexity, for example if you want to re-create the future with a different File when the socket receives a message.

And indeed, the linked fix there is to pin the future, which is exactly what we were doing.

So...dunno.  I guess I'd probably still vote for merging this.  The code is perhaps a bit clearer at least, and `select!` is still a trap - but ISTM that a linter could warn about any "nontrivial" non-pinned futures somehow or so?

Reply:
I think both versions are correct, but as the blogpost mentions this kind of code is hard to reason about and review.
I'm happy to land this new macro-less version.